.github/workflows/pull-request.yaml

```diff
@@ -25,78 +25,87 @@ jobs:
           cargo clippy --version
       - run: cargo +nightly fmt --all -- --check
       - run: cargo clippy --locked --workspace --all-features --all-targets -- -D warnings
-  rust-tests:
+
+  unit-tests:
     timeout-minutes: 60
     runs-on: ubuntu-latest
     env:
       # Shrink artifact size by not including debug info. Makes build faster and shrinks cache.
       CARGO_PROFILE_DEV_DEBUG: 0
       CARGO_PROFILE_TEST_DEBUG: 0
-      FLYWAY_VERSION: 7.3.1
-      FLYWAY_PATH: /home/runner/flyway
-      # Flyway is connecting via TCP/IP (authentication via password), whereas psql connects via Unix-domain sockets (authentication via username)
-      # We set the db password of $USER to this value which is also used by the flyway command
-      FLYWAY_PASSWORD: password
-      # Extra underscore because conflicts with hardhat internally used variables.
-      HARDHAT_VERSION_: 2.6.8
-      HARDHAT_PATH_: /home/runner/hardhat
     steps:
       - uses: actions/checkout@v2
       - uses: Swatinem/rust-cache@v1
-      - name: Cache flyway
-        id: cache-flyway
-        uses: actions/cache@v2
+      - run: cargo test
+
+  test-db:
+    timeout-minutes: 60
+    runs-on: ubuntu-latest
+    env:
+      # Shrink artifact size by not including debug info. Makes build faster and shrinks cache.
+      CARGO_PROFILE_DEV_DEBUG: 0
+      CARGO_PROFILE_TEST_DEBUG: 0
+    steps:
+      - uses: actions/checkout@v2
+      - uses: Swatinem/rust-cache@v1
+      - uses: actions-rs/toolchain@v1
         with:
-          path: ${{ env.FLYWAY_PATH }}
-          key: ${{ runner.os }}-flyway-${{ env.FLYWAY_VERSION }}
-      - name: Install Flyway
-        if: steps.cache-flyway.outputs.cache-hit != 'true'
-        run: |
-          mkdir --parents ${{ env.FLYWAY_PATH }}
-          cd ${{ env.FLYWAY_PATH }}
-          curl -L https://repo1.maven.org/maven2/org/flywaydb/flyway-commandline/${FLYWAY_VERSION}/flyway-commandline-${FLYWAY_VERSION}.tar.gz -o out.tar.gz
-          tar --strip-components=1 -xzf out.tar.gz
-          rm out.tar.gz
-          ls -al
-      - name: Cache Hardhat
-        id: cache-hardhat
-        uses: actions/cache@v2
+          profile: minimal
+          toolchain: nightly
+          components: rustfmt
+      # Start the build process in the background. The following cargo test command will automatically
+      # wait for the build process to be done before proceeding.
+      - run: cargo build -p orderbook -p database --tests &
```

Comment:
Would it be possible to only build all the targets once and still run the tests in parallel afterwards?
Maybe we can create a docker image with the built binaries in one job and let other jobs use that image to run the tests?

This should also get rid of a lot of duplicated setup code.

Reply:
So, it is possible in theory by using [workflow artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts), and I did try this. However, it's just another instance of [software](https://github.com/actions/upload-artifact/issues/199), so no go :(